File: backend/core/parser.py
from core.lexer import TokenType
from core.tree import BinaryOp, Equation, Number, Variable


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def current(self):
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def eat(self, token_type=None):
        token = self.current()
        if token is None:
            return None

        if token_type and token.type != token_type:
            raise Exception(f"Expected {token_type}, got {token.type}")

        self.pos += 1
        return token

    def parse(self):
        left = self.expr()

        # 🔥 FIX: make sure '=' is detected correctly
        if self.current() and self.current().type == TokenType.EQUAL:
            self.eat(TokenType.EQUAL)
            right = self.expr()
            eq = Equation(left, right)

            # convert to solver format
            return self.build_equation(eq)

        # ❌ NEVER return raw AST here
        raise Exception("Invalid equation format. '=' missing or not parsed correctly")

    def expr(self):
        node = self.term()

        while self.current() and self.current().type in (TokenType.PLUS, TokenType.MINUS):
            op = self.eat().type
            right = self.term()
            node = BinaryOp(node, op, right)

        return node

    def term(self):
        node = self.factor()

        while True:
            curr = self.current()

            # explicit * or /
            if curr and curr.type in (TokenType.MUL, TokenType.DIV):
                op = self.eat().type
                right = self.factor()
                node = BinaryOp(node, op, right)

            # 🔥 implicit multiplication (IMPORTANT FIX)
            elif curr and curr.type in (TokenType.NUMBER, TokenType.VARIABLE, TokenType.LPAREN):
                # treat as multiplication
                right = self.factor()
                node = BinaryOp(node, TokenType.MUL, right)

            else:
                break

        return node

    def factor(self):
        token = self.current()

        if token.type == TokenType.NUMBER:
            self.eat(TokenType.NUMBER)
            return Number(token.value)

        if token.type == TokenType.VARIABLE:
            self.eat(TokenType.VARIABLE)
            return Variable(token.value)

        if token.type == TokenType.LPAREN:
            self.eat(TokenType.LPAREN)
            node = self.expr()
            self.eat(TokenType.RPAREN)
            return node

        raise Exception(f"Unexpected token: {token}")

    # -------------------------
    # 🔥 IMPORTANT PART
    # Convert AST → coefficients
    # -------------------------
    def build_equation(self, eq):
        left_coeffs, left_const = self.eval_poly(eq.left)
        right_coeffs, right_const = self.eval_poly(eq.right)

        # Move all variables to left, all constants to right
        coeffs = {}
        for k, v in left_coeffs.items():
            coeffs[k] = coeffs.get(k, 0) + v
        for k, v in right_coeffs.items():
            coeffs[k] = coeffs.get(k, 0) - v

        target = right_const - left_const

        return {
            "coefficients": coeffs,
            "target": target
        }

    def eval_poly(self, node):
        if isinstance(node, Number):
            return {}, node.value

        if isinstance(node, Variable):
            return {node.name: 1}, 0

        if isinstance(node, BinaryOp):
            l_coeffs, l_const = self.eval_poly(node.left)
            r_coeffs, r_const = self.eval_poly(node.right)

            if node.op == TokenType.PLUS:
                c = dict(l_coeffs)
                for k, v in r_coeffs.items():
                    c[k] = c.get(k, 0) + v
                return c, l_const + r_const

            if node.op == TokenType.MINUS:
                c = dict(l_coeffs)
                for k, v in r_coeffs.items():
                    c[k] = c.get(k, 0) - v
                return c, l_const - r_const

            if node.op == TokenType.MUL:
                if l_coeffs and r_coeffs:
                    raise Exception("Non-linear equations (variable * variable) are not supported.")

                c = {}
                if l_coeffs:  # left has variables, so multiply them by right constant
                    for k, v in l_coeffs.items():
                        c[k] = v * r_const
                    return c, l_const * r_const
                else:  # right has variables, so multiply them by left constant
                    for k, v in r_coeffs.items():
                        c[k] = v * l_const
                    return c, l_const * r_const

            if node.op == TokenType.DIV:
                if r_coeffs:
                    raise Exception("Dividing by a variable in the equation is not supported.")
                if r_const == 0:
                    raise Exception("Division by zero.")

                c = {}
                for k, v in l_coeffs.items():
                    if v % r_const != 0:
                        raise Exception("Fractional coefficients are not supported.")
                    c[k] = v // r_const
                if l_const % r_const != 0:
                    raise Exception("Fractional constants are not supported.")
                return c, l_const // r_const

        raise Exception(f"Unrecognized node in expression: {type(node)}")

File: backend/core/parser.cpp
#include "parser.h"
#include <stdexcept>

namespace quant {

static Expression makeConstant(int value) {
    Expression expr;
    expr.constant = value;
    return expr;
}

static Expression makeVariable(const std::string &name, int coefficient = 1) {
    Expression expr;
    expr.coefficients[name] = coefficient;
    return expr;
}

Expression Parser::add(const Expression &lhs, const Expression &rhs) {
    Expression result = lhs;
    result.constant += rhs.constant;
    for (const auto &entry : rhs.coefficients) {
        result.coefficients[entry.first] += entry.second;
        if (result.coefficients[entry.first] == 0) {
            result.coefficients.erase(entry.first);
        }
    }
    return result;
}

Expression Parser::subtract(const Expression &lhs, const Expression &rhs) {
    Expression result = lhs;
    result.constant -= rhs.constant;
    for (const auto &entry : rhs.coefficients) {
        result.coefficients[entry.first] -= entry.second;
        if (result.coefficients[entry.first] == 0) {
            result.coefficients.erase(entry.first);
        }
    }
    return result;
}

Expression Parser::multiply(const Expression &lhs, const Expression &rhs) {
    bool lhsHasVars = !lhs.coefficients.empty();
    bool rhsHasVars = !rhs.coefficients.empty();
    if (lhsHasVars && rhsHasVars) {
        throw std::invalid_argument("Multiplication of two variables is not supported.");
    }

    if (lhsHasVars) {
        int scalar = rhs.constant;
        if (scalar == 0) {
            return makeConstant(0);
        }
        Expression result;
        result.constant = lhs.constant * scalar;
        for (const auto &entry : lhs.coefficients) {
            result.coefficients[entry.first] = entry.second * scalar;
        }
        return result;
    }

    if (rhsHasVars) {
        int scalar = lhs.constant;
        if (scalar == 0) {
            return makeConstant(0);
        }
        Expression result;
        result.constant = rhs.constant * scalar;
        for (const auto &entry : rhs.coefficients) {
            result.coefficients[entry.first] = entry.second * scalar;
        }
        return result;
    }

    return makeConstant(lhs.constant * rhs.constant);
}

Expression Parser::divide(const Expression &lhs, const Expression &rhs) {
    if (!rhs.coefficients.empty()) {
        throw std::invalid_argument("Division by a variable expression is not supported.");
    }
    if (rhs.constant == 0) {
        throw std::invalid_argument("Division by zero is not allowed.");
    }
    if (lhs.constant % rhs.constant != 0) {
        throw std::invalid_argument("Expression must evaluate to whole numbers only.");
    }
    Expression result;
    result.constant = lhs.constant / rhs.constant;
    for (const auto &entry : lhs.coefficients) {
        if (entry.second % rhs.constant != 0) {
            throw std::invalid_argument("Expression must evaluate to whole numbers only.");
        }
        result.coefficients[entry.first] = entry.second / rhs.constant;
    }
    return result;
}

ParseResult Parser::parseEquation(const std::string &expression) {
    Lexer lexer(expression);
    auto tokens = lexer.tokenize();

    size_t equalsIndex = tokens.size();
    for (size_t i = 0; i < tokens.size(); ++i) {
        if (tokens[i].type == TokenType::Equals) {
            equalsIndex = i;
            break;
        }
    }

    if (equalsIndex == tokens.size()) {
        throw std::invalid_argument("Expression must contain '=' to separate sides.");
    }

    size_t leftIndex = 0;
    Expression leftExpr = parseExpression(tokens, leftIndex, equalsIndex);
    if (leftIndex != equalsIndex) {
        throw std::invalid_argument("Invalid expression on left side of '='.");
    }

    size_t rightIndex = equalsIndex + 1;
    Expression rightExpr = parseExpression(tokens, rightIndex, tokens.size());
    if (rightIndex != tokens.size()) {
        throw std::invalid_argument("Invalid expression on right side of '='.");
    }

    ParseResult result;
    result.coefficients = leftExpr.coefficients;
    for (const auto &entry : rightExpr.coefficients) {
        result.coefficients[entry.first] -= entry.second;
    }
    result.target = rightExpr.constant - leftExpr.constant;

    for (auto it = result.coefficients.begin(); it != result.coefficients.end();) {
        if (it->second == 0) {
            it = result.coefficients.erase(it);
        } else {
            ++it;
        }
    }

    return result;
}

Expression Parser::parseExpression(const std::vector<Token> &tokens, size_t &index, size_t end) {
    Expression result = parseTerm(tokens, index, end);
    while (index < end && (tokens[index].type == TokenType::Plus || tokens[index].type == TokenType::Minus)) {
        TokenType op = tokens[index].type;
        ++index;
        Expression rhs = parseTerm(tokens, index, end);
        result = (op == TokenType::Plus) ? add(result, rhs) : subtract(result, rhs);
    }
    return result;
}

Expression Parser::parseTerm(const std::vector<Token> &tokens, size_t &index, size_t end) {
    Expression result = parseFactor(tokens, index, end);
    while (index < end && (tokens[index].type == TokenType::Multiply || tokens[index].type == TokenType::Divide)) {
        TokenType op = tokens[index].type;
        ++index;
        Expression rhs = parseFactor(tokens, index, end);
        result = (op == TokenType::Multiply) ? multiply(result, rhs) : divide(result, rhs);
    }
    return result;
}

Expression Parser::parseFactor(const std::vector<Token> &tokens, size_t &index, size_t end) {
    if (index >= end) {
        throw std::invalid_argument("Unexpected end of expression.");
    }

    if (tokens[index].type == TokenType::Plus || tokens[index].type == TokenType::Minus) {
        TokenType sign = tokens[index].type;
        ++index;
        Expression factor = parseFactor(tokens, index, end);
        if (sign == TokenType::Minus) {
            for (auto &entry : factor.coefficients) {
                entry.second = -entry.second;
            }
            factor.constant = -factor.constant;
        }
        return factor;
    }

    if (tokens[index].type == TokenType::Number) {
        int value = static_cast<int>(tokens[index].value);
        ++index;
        if (index < end && tokens[index].type == TokenType::Variable) {
            std::string variable = tokens[index].text;
            ++index;
            return makeVariable(variable, value);
        }
        return makeConstant(value);
    }

    if (tokens[index].type == TokenType::Variable) {
        std::string variable = tokens[index].text;
        ++index;
        return makeVariable(variable, 1);
    }

    if (tokens[index].type == TokenType::LeftParen) {
        ++index;
        Expression inner = parseExpression(tokens, index, end);
        if (index >= end || tokens[index].type != TokenType::RightParen) {
            throw std::invalid_argument("Mismatched parentheses in expression.");
        }
        ++index;
        return inner;
    }

    throw std::invalid_argument("Unexpected token in expression.");
}

} // namespace quant
